src/libraries/Microsoft.Extensions.Http/src/DependencyInjection/HttpClientBuilderExtensions.Logging.cs

```diff
@@ -0,0 +1,147 @@
+        /// <param name="httpClientLoggerFactory">A delegate that is used to create a custom logger. The logger should implement
+        /// <see cref="IHttpClientLogger"/> or <see cref="IHttpClientAsyncLogger"/>.</param>
+        /// <param name="wrapHandlersPipeline">Whether the logging handler with the custom logger would be added to the top
+        /// or to the bottom of the additional handlers chains.</param>
+        /// <returns>An <see cref="IHttpClientBuilder"/> that can be used to configure the client.</returns>
+        /// <remarks>
+        /// <para>
+        /// If the <see paramref="wrapHandlersPipeline"/> is `true`, <see cref="IHttpClientLogger.LogRequestStart"/> and
+        /// <see cref="IHttpClientAsyncLogger.LogRequestStartAsync"/> would be executed before all
+        /// other additional handlers in the chain. <see cref="IHttpClientLogger.LogRequestStop"/> and
+        /// <see cref="IHttpClientAsyncLogger.LogRequestStopAsync"/> would be executed after all
+        /// other additional handlers, essentially wrapping the whole pipeline.
+        /// </para>
+        /// <para>
+        /// If the <see paramref="wrapHandlersPipeline"/> is `false`, <see cref="IHttpClientLogger.LogRequestStart"/> and
+        /// <see cref="IHttpClientAsyncLogger.LogRequestStartAsync"/> would be executed after all
+        /// other additional handlers in the chain, right before the primary handler. <see cref="IHttpClientLogger.LogRequestStop"/> and
+        /// <see cref="IHttpClientAsyncLogger.LogRequestStopAsync"/> would be executed before all
+        /// other additional handlers, right after the primary handler.
+        /// </para>
+        /// <para>
+        /// The <see cref="IServiceProvider"/> argument provided to <paramref name="httpClientLoggerFactory"/> will be
+        /// a reference to a scoped service provider that shares the lifetime of the handler chain being constructed.
+        /// </para>
+        /// <para>
+        /// If <see cref="AddLogger"/> is called multiple times, multiple loggers would be added. If <see cref="RemoveAllLoggers"/> was
+        /// not called before calling <see cref="AddLogger"/>, then new logger would be added in addition to the default ones.
+        /// </para>
+        /// </remarks>
+        public static IHttpClientBuilder AddLogger(this IHttpClientBuilder builder, Func<IServiceProvider, IHttpClientLogger> httpClientLoggerFactory, bool wrapHandlersPipeline = false)
+        {
+            ThrowHelper.ThrowIfNull(builder);
+            ThrowHelper.ThrowIfNull(httpClientLoggerFactory);
+
+            builder.Services.Configure<HttpClientFactoryOptions>(builder.Name, options =>
+            {
+                options.LoggingBuilderActions.Add(b =>
+                {
+                    IHttpClientLogger httpClientLogger = httpClientLoggerFactory(b.Services);
+                    HttpClientLoggerHandler handler = new HttpClientLoggerHandler(httpClientLogger);
```

Comment:
Is the expectation that there will be very few such loggers (0 or 1) for a given client?

Reply:
I think so? I don't see why anyone would need more than 2 (in the beginning and in the end of the handlers chain)